### PR TITLE
refactor: combine multiple trace tests into single file

### LINKED ISSUE #168: Combine multiple trace tests into single file
When one runs `time atomkraft test trace --path <dir-path>/ ...` where `<dir-path>` is a directory that contains many trace files, Atomkraft generates one python test file for each trace. A test file is simple 3 lines of code:
```
@itf(<path-to-itf-json-file>, keypath=<keypath>)
def test_trace():
    logging.info("Successfully executed trace " + <path-to-itf-json-file>)
```

A better approach would be to combine all tests in each subdirectory of `<dir-path>` into a single file:
```
@itf(<path-to-itf-json-file1>, keypath=<keypath>)
def test_<path-to-itf-json-file1>():
    logging.info("Successfully executed trace " + <path-to-itf-json-file1>)

@itf(<path-to-itf-json-file2>, keypath=<keypath>)
def test_<path-to-itf-json-file2>:
    logging.info("Successfully executed trace " + <path-to-itf-json-file2>)

```
This would allow the user to re-execute all the traces in a subdirectory with a single `pytest` command.

### PR DESCRIPTION
**Pull request by hvanz**
Closes #168 

- `test trace` on a directory will create a single pytest file with test cases for all traces, instead of one file per trace.
- Refactor `test trace` and `test model`, as there was a lot of duplicate code.

### CODE REVIEW

File: atomkraft/utils/filesystem.py [hunk 1]
```diff
@@ -7,3 +7,7 @@ def last_modified_file_in(dir: Path) -> Path:
     """
     all_files = Path(dir).glob(f"{dir}/*")
     return max(all_files, key=lambda x: x.stat().st_mtime)
+
+
+def rename_chars(path: str) -> str:
+    return path.replace("/", "_").replace(".", "_").replace(":", "_").replace("-", "_")
```
Comment (rnbguy) [inline atomkraft/utils/filesystem.py]: Can we use [`case-converter`](https://pypi.org/project/case-converter) instead?

File: atomkraft/utils/helpers.py [hunk 2]
```diff
@@ -1,7 +1,15 @@
...
 # removes the suffix `suffix` from `text`.
 # Implemented here as a helper fucntion in order to support Python 3.8
 def remove_suffix(text: str, suffix: str):
     if text.endswith(suffix):
         text = text[: -len(suffix)]
 
     return text
+
+
+def natural_sort(key):
```
Comment (rnbguy) [inline atomkraft/utils/helpers.py]: Please annotate with type `pathlib.Path`.

File: atomkraft/utils/helpers.py [hunk 3]
```diff
@@ -1,7 +1,15 @@
...
+    """For sorting strings containing numbers without leading zeroes"""
+    return [int(s) if s.isdigit() else s.lower() for s in re.split("(\\d+)", str(key))]
```
Comment (rnbguy) [inline atomkraft/utils/helpers.py:24]: ~I am confused by `str(key)` here. Can you give me an example of non-`str` key for this method?~

This is a method to compute keys for a sort. Does it make sense to rename the method to `natural_key`?

File: atomkraft/test/model.py [hunk 4]
```diff
@@ -77,14 +78,7 @@ def test_model(
...
 
-    test_group = f"{model.stem}_{timestamp}"
-    test_group = (
-        test_group.replace("/", "_")
-        .replace(".", "_")
-        .replace(":", "_")
-        .replace("-", "_")
-    )
-
+    test_group = rename_chars(f"{model.stem}_{timestamp}")
```
Comment (rnbguy) [inline atomkraft/test/model.py]: thanks for creating the util method for this! :+1:

**Review by rnbguy — CHANGES_REQUESTED** [review]
Thanks for the PR. I added some suggestions.

I noticed, this is only for `test trace`. `test model` still generates multiple test files. Do you think it would make sense to include that in this PR as well?

File: atomkraft/utils/filesystem.py (hunk 1 above)
Comment (hvanz) [reply atomkraft/utils/filesystem.py]: Done.

**Comment by hvanz** [comment]
> I noticed, this is only for `test trace`. `test model` still generates multiple test files. Do you think it would make sense to include that in this PR as well?

@rnbguy I included this change you suggested in this PR, together with a refactoring of all the testing functionality in a single `Test` class

File: atomkraft/test/test.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+TEST_FILE_TEST_TRACE_STUB = """
+
+@itf({0}, keypath={2})
+def test_{1}():
+    logging.info("Successfully executed trace " + {0})
+"""
+
+
+class Test:
```
Comment (rnbguy) [inline atomkraft/test/test.py:33]: Great that you used a class for this.

File: atomkraft/utils/filesystem.py [hunk 6]
```diff
@@ -7,3 +11,21 @@ def last_modified_file_in(dir: Path) -> Path:
     """
     all_files = Path(dir).glob(f"{dir}/*")
     return max(all_files, key=lambda x: x.stat().st_mtime)
+
+
+def snakecase(path: str) -> str:
+    return caseconverter.snakecase(path, delimiters="/.:-")
+
+
+def copy_if_exists(srcs: Union[Path, List[Path]], dst_path: Path):
```
Comment (rnbguy) [inline atomkraft/utils/filesystem.py:20]: Thanks for putting this here.

**Review by rnbguy — APPROVED** [review]
Awesome ! Thanks for this PR.

**Pull request merged**